portalnet/src/gossip.rs

```diff
@@ -43,80 +39,68 @@ pub struct GossipResult {
 
 /// Propagate gossip in a way that can be used across threads, without &self.
 /// Doesn't trace gossip results
-pub fn propagate_gossip_cross_thread<TContentKey: OverlayContentKey>(
+pub fn propagate_gossip_cross_thread<TContentKey: OverlayContentKey, TMetric: Metric>(
     content: Vec<(TContentKey, Vec<u8>)>,
     kbuckets: Arc<RwLock<KBucketsTable<NodeId, Node>>>,
     command_tx: mpsc::UnboundedSender<OverlayCommand<TContentKey>>,
     utp_controller: Option<Arc<UtpController>>,
 ) -> usize {
-    // Propagate all validated content, whether or not it was stored.
-    let ids_to_propagate: Vec<String> = content
-        .iter()
-        .unique_by(|(key, _)| key.content_id())
-        .map(|(key, _)| hex_encode_compact(key.content_id()))
-        .collect();
-    debug!(ids = ?ids_to_propagate, "propagating validated content");
-
-    // Get all connected nodes from overlay routing table
-    let kbuckets = kbuckets.read();
-    let all_nodes: Vec<&kbucket::Node<NodeId, Node>> = kbuckets
-        .buckets_iter()
-        .flat_map(|kbucket| {
-            kbucket
-                .iter()
-                .filter(|node| node.status.is_connected())
-                .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+    // Precalculate content ids
+    let content = content
+        .into_iter()
+        .map(|(content_key, content_value)| {
+            (content_key.content_id(), (content_key, content_value))
         })
-        .collect();
+        .collect::<HashMap<_, _>>();
 
+    let content_ids = content.keys().collect::<Vec<_>>();
     debug!(
-        "propagating validated content: found {} nodes",
-        all_nodes.len()
+        ids = ?content_ids.iter().map(hex_encode_compact),
+        "propagating validated content",
     );
-    if all_nodes.is_empty() {
-        // If there are no nodes whatsoever in the routing table the gossip cannot proceed.
-        warn!("No nodes in routing table, gossip cannot proceed.");
-        return 0;
-    }
 
-    // HashMap to temporarily store all interested ENRs and the content.
-    // Key is base64 string of node's ENR.
-    let mut enrs_and_content: HashMap<String, Vec<(TContentKey, Vec<u8>)>> = HashMap::new();
+    // Map from content_ids to interested ENRs
+    let mut interested_ends = batch_interested_enrs::<TMetric>(&content_ids, kbuckets);
```

Comment:
`interested_enrs`

Reply:
Or I guess 
```suggestion
    let mut interested_enr_batch = batch_interested_enrs::<TMetric>(&content_ids, kbuckets);
```
Since `interested_enrs` is used below for those interested in a single ID.